File: JAVA/java/src/SmallestStringStartingFromLeaf.java
import java.util.*;

class TreeNode7 {
    int val;
    TreeNode7 left, right;

    TreeNode7(int val) {
        this.val = val;
        left = null;
        right = null;
    }
}

public class SmallestStringStartingFromLeaf {
    public String smallestFromLeaf(TreeNode7 root) {
        if (root == null)
            return "";

        String smallestString = "";
        Queue<Map.Entry<TreeNode7, String>> nodeQueue = new LinkedList<>();
        nodeQueue.offer(new AbstractMap.SimpleEntry<>(root, String.valueOf((char) (root.val + 'a'))));

        while (!nodeQueue.isEmpty()) {
            Map.Entry<TreeNode7, String> entry = nodeQueue.poll();
            TreeNode7 node = entry.getKey();
            String currentString = entry.getValue();

            if (node.left == null && node.right == null) {
                if (smallestString.isEmpty() || currentString.compareTo(smallestString) < 0)
                    smallestString = currentString;
            }

            if (node.left != null)
                nodeQueue.offer(new AbstractMap.SimpleEntry<>(node.left, (char) (node.left.val + 'a') + currentString));

            if (node.right != null)
                nodeQueue.offer(new AbstractMap.SimpleEntry<>(node.right, (char) (node.right.val + 'a') + currentString));
        }

        return smallestString;
    }

    public static TreeNode7 arrayToTree(Integer[] arr) {
        if (arr == null || arr.length == 0 || arr[0] == null)
            return null;

        TreeNode7 root = new TreeNode7(arr[0]);
        Queue<TreeNode7> queue = new LinkedList<>();
        queue.offer(root);
        int i = 1;

        while (!queue.isEmpty() && i < arr.length) {
            TreeNode7 currNode = queue.poll();

            if (arr[i] != null) {
                currNode.left = new TreeNode7(arr[i]);
                queue.offer(currNode.left);
            }
            i++;

            if (i < arr.length && arr[i] != null) {
                currNode.right = new TreeNode7(arr[i]);
                queue.offer(currNode.right);
            }
            i++;
        }
        return root;
    }

    public static void main(String[] args) {
        Integer[] nums = {0, 1, 2, 3, 4, 3, 4};
        TreeNode7 root = arrayToTree(nums);
        SmallestStringStartingFromLeaf solution = new SmallestStringStartingFromLeaf();
        System.out.println(solution.smallestFromLeaf(root));
    }
}

File: python/Smallest String Starting From Leaf.py
from collections import deque
from typing import Optional

# TreeNode definition
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def list_to_tree(nums):
    if not nums:
        return None

    root = TreeNode(nums[0])
    queue = deque([root])
    i = 1

    while i < len(nums):
        node = queue.popleft()

        if nums[i] is not None:
            node.left = TreeNode(nums[i])
            queue.append(node.left)
        i += 1

        if i < len(nums) and nums[i] is not None:
            node.right = TreeNode(nums[i])
            queue.append(node.right)
        i += 1

    return root

def smallestFromLeaf(root: Optional[TreeNode]) -> str:
    if not root:
        return ""

    smallest_string = ""
    node_queue = deque()
    node_queue.append((root, chr(root.val + ord('a'))))

    while node_queue:
        node, current_string = node_queue.popleft()

        if not node.left and not node.right:
            smallest_string = min(smallest_string, current_string) if smallest_string else current_string

        if node.left:
            node_queue.append((node.left, chr(node.left.val + ord('a')) + current_string))

        if node.right:
            node_queue.append((node.right, chr(node.right.val + ord('a')) + current_string))

    return smallest_string

# Example usage:
nums = [0,1,2,3,4,3,4]
tree = list_to_tree(nums)
print(smallestFromLeaf(tree))
'''
You are given the root of a binary tree where each node has a value in the range [0, 25] representing the letters 'a' to 'z'.

Return the lexicographically smallest string that starts at a leaf of this tree and ends at the root.

As a reminder, any shorter prefix of a string is lexicographically smaller.

For example, "ab" is lexicographically smaller than "aba".
A leaf of a node is a node that has no children.



Example 1:


Input: root = [0,1,2,3,4,3,4]
Output: "dba"
Example 2:


Input: root = [25,1,3,1,3,0,2]
Output: "adz"
Example 3:


Input: root = [2,2,1,null,1,0,null,0]
Output: "abc"
'''